### PR TITLE
Make task spawning for method call optional

### PR DESCRIPTION
**Pull request by elmarco**
Clients can expect calls to be handled in order they are sent. Although the specification doesn't explicitly state so  (https://gitlab.freedesktop.org/dbus/dbus/-/issues/179), bus implementations have respected message ordering. This can also be  expected when the bus and client are fully managed, as in p2p.

Add an option to avoid the default task spawning when dispatching methods on a specific interface.

This currently prevents parallel handling of other incoming calls which will have to wait for the method to return (current pending calls are still running in parallel).

Supersedes: #668

### CODE REVIEW

**Comment by zeenix** [comment]
> Clients can expect calls to be handled in order they are sent. Although the specification doesn't explicitly state so (https://gitlab.freedesktop.org/dbus/dbus/-/issues/179), 

As explained multiple times, the spec does not say it at all in any way or form, not even implicitly or remotely hinting towards it. Please stop insisting that it's the case, unless/until you get a confirmation from any of the spec maintainers/authors on your assertion here.

> bus implementations have respected message ordering.

Order of messages for the bus is important because bus needs to replicate the behavior of the service. Very strange things can happen under very normal circumstances if bus reorders messages (no batching or strange assumptions from clients needed). Hence I find this very irrelevant.

> this can also be expected when the bus and client are fully managed, as in p2p.

This is the only valid use case (even though it's even more niche since its only clients that batch method calls with inter-dependencies) here. I has to ask you to modify the description of the PR and the relevant commit message to focus solely on this use case.

I'll review once you've done that.

File: zbus_macros/src/lib.rs
Comment (zeenix) [inline zbus_macros/src/lib.rs:225]: * Remove "currently" and update to reflect that we're talking about method handling order. Message ordering ordering and this linked issue is very much irrelevant.
* I don't think this should be decided on per interfaces bases. I just want a simple boolean setting on either ConnectionBuilder or ObjectServer.

**Comment by danieldg** [comment]
Another possible use case that you could document is inter-operating with an existing client that assumes this type of ordering (including when it does it unintentionally, or by using `NoReplyExpected`).  Accidental dependencies on things like this (which turn in to race conditions) are unfortunately common and hard to debug, so identifying this as a "make buggy clients happy" knob could help - both in making things work and in pointing out that it is in fact a bug in the client, not zbus.

File: zbus_macros/src/lib.rs
Comment (zeenix) [reply zbus_macros/src/lib.rs:225]: @elmarco actually I've a better suggestion that you'll probably like: we can make non-spawning the default for p2p. Since separate clients will be on separate connections anyway and hence have their ObjectServer running in a separate task. 

Also we don't need to provide an option to change this behaviour unless someone asks for it. I don't see much point in parallelising method calls from the same client. We should document this behaviour though.

I'd be fine with doing the same for the bus case but that would be adding a lot of complexity to track all the peers and managing tasks for them.

Comment (elmarco) [reply zbus_macros/src/lib.rs:225]: > @elmarco actually I've a better suggestion that you'll probably like: we can make non-spawning the default for p2p. Since separate clients will be on separate connections anyway and hence have their ObjectServer running in a separate task.

We could discuss the various use cases having an option to set the spawning behaviour at either Method, Interface, ObjectServer, Connection. They are *all* valid to me.  

But I think the sweet spot is at the Interface level, because it is a set of related methods and implementation, and you can document this expected behaviour better than at the ObjectServer or Connection level. At the ObjectServer or Connection level, you enforce a behaviour on other implementations that may not require it.

There is no fundamental difference from a p2p connection and bus connection, and both could serve a single interface or multiple interface/multiplexed communication. Imho, we shouldn't behave differently because of that.

> Also we don't need to provide an option to change this behaviour unless someone asks for it. I don't see much point in parallelising method calls from the same client. We should document this behaviour though.

There is as much reasons to want or no want parallelism with p2p or bus connection.

**Comment by zeenix** [comment]
BTW, if you can break the cleanup/fixes commits into a separate PR, we can already merge those.

File: zbus_macros/src/lib.rs
Comment (elmarco) [reply zbus_macros/src/lib.rs:225]: > > There is no fundamental difference from a p2p connection and bus connection, and both could serve a single interface or multiple interface/multiplexed communication.
> 
> It's not about the number of interfaces served. It's about the relationship between client and connection 

The same p2p connection can be shared for different usages & interfaces, just like a bus connection.

> 
> > Imho, we shouldn't behave differently because of that.
> 
> The only use-case you presented was about p2p and if we can satisfy that use case w/o needing new API (at least temporarily), I'd prefer that.

It's not specific to p2p. org.qemu.Display1.Keyboard has Press and Release methods for example. There is no reason a client would have to wait for a reply for Press before sending Release. This interface is exposed on the default bus by default. The implementation is currently C/GDBus, but you should get my point.

Comment (zeenix) [reply zbus_macros/src/lib.rs:225]: > > > There is no fundamental difference from a p2p connection and bus connection, and both could serve a single interface or multiple interface/multiplexed communication.
> > 
> > 
> > It's not about the number of interfaces served. It's about the relationship between client and connection
> 
> The same p2p connection can be shared for different usages & interfaces, just like a bus connection

As I wrote, it's not about interfaces (only if we take the approach in this PR, which I'm objecting to for other reasons I mentioned). 

> It's not specific to p2p. org.qemu.Display1.Keyboard has Press and Release methods for example. 

Again, dependencies between methods is **not** an argument alone. You can find countless examples of methods depending on each other out in the wild.

> There is no reason a client would have to wait for a reply for Press before sending Release.

If it's from the same client, yes they shouldn't need to. Ideally both in case of p2p and bus, but as I wrote, it'd be hard to implement it for the bus but in case of p2p, there is only one client per connection so it's easy to make it happen.

Comment (elmarco) [reply zbus_macros/src/lib.rs:225]: > > The same p2p connection can be shared for different usages & interfaces, just like a bus connection
> 
> As I wrote, it's not about interfaces (only if we take the approach in this PR, which I'm objecting to for other reasons I mentioned).

I disagree, typically a set of related methods on an interface must be handled in order. If you are lucky enough to have an interface that doesn't need ordering, you can keep zbus default behaviour.

p2p or bus connection, it doesn't matter, they can serve several unrelated objects & interfaces with different constraints. Your point about making several p2p connections doesn't hold, they are usually sharing the same connection to uniquely identify a client and avoid pointless extra resources usage and connection setups.

> 
> > It's not specific to p2p. org.qemu.Display1.Keyboard has Press and Release methods for example.
> 
> Again, dependencies between methods is **not** an argument alone. You can find countless examples of methods depending on each other out in the wild.

Good you agree they are countless existing examples where ordering or methods matters!

Comment (zeenix) [reply zbus_macros/src/lib.rs:225]: > Good you agree they are countless existing examples where ordering of methods matters! (the norm, not the exception to me)

I never disagreed with that at all. If you think I did, you misunderstood. A method call in general is incomplete w/o a return. If you had 2 async methods with interdependency, you won't do something like this in Rust:

```rust
// `method_b` depends on the side effects of `method_a` here.
let batch = [method_a(), method_b()];
...
for method in batch {
    tokio::task::spawn(method);
}
```

---

Since it's clear to me that we won't be able to convince each other after having spent a lot of our time on this already, I am no longer interested in discussing it further. I hate to do this but I'm pulling the maintainer card and ask you to either take one of the 2 options I suggested (either make non-spawning default only for the p2p and document this behavior, or add the ConnectionBuilder/ObjectServer method to disable task spawning) or close the PR.

Comment (elmarco) [reply zbus_macros/src/lib.rs:225]: With task spawning, it's quite explicit that there are no dependencies. The trouble is the client is simply doing `send([methodA,methodB])` and can reasonably expect that to be handled in order, unless the implementation clearly state otherwise.

Comment (elmarco) [reply zbus_macros/src/lib.rs:225]: Setting a no-spawn option _globally_ on Connection/ObjectServer brings back the potential dead-lock issue https://github.com/dbus2/zbus/commit/2da0ecd98f449aa2fcae2ff7f656fe91a13d8b9a. By setting `no_spawn` on the Interface, the _specific_ implementation states that it doesn't rely on reentering (no blocking calls etc).

Comment (zeenix) [reply zbus_macros/src/lib.rs:225]: > Setting a no-spawn option _globally_ on Connection/ObjectServer brings back the potential dead-lock issue [2da0ecd](https://github.com/dbus2/zbus/commit/2da0ecd98f449aa2fcae2ff7f656fe91a13d8b9a)

I don't think so. The commit was made before ObjectServer started using its own MessageStream (i-e queue), which is separate from the stream used by method replies so the deadlock mentioned in the commit log is not possible anymore. Almost everything uses its own stream now in zbus even and this was done to avoid all such deadlocks. 

But maybe I'm missing something. Can describe where and how this can deadlock?

Comment (danieldg) [reply zbus_macros/src/lib.rs:225]: The stream is still bounded (by `DEFAULT_MAX_QUEUED`). Since it's bounded, you could get this deadlock if you wait on any dbus recv operation inside your handler (the usual cause of such a wait would be "making a dbus call").  And even if it were unbounded (or the bound is not hit), you can still get a deadlock if you have a circular dependency of dbus calls that cycles back to this ObjectServer.  The circular dependency is a bit easier to just document as "don't risk doing that if you set no-spawn".

Comment (elmarco) [reply zbus_macros/src/lib.rs:225]: @danieldg +1, correct

Comment (zeenix) [reply zbus_macros/src/lib.rs:225]: > Since it's bounded, you could get this deadlock if you wait on any dbus recv operation inside your handler (the usual cause of such a wait would be "making a dbus call").

True. However, although after all the debates on this topic, I still remain unconvinced that anyone other than @elmarco needs this anyway but just in case I'm wrong and there are other people/projects out there "batching method calls with inter-dependencies", I am open to a connection option and I think it's more than sufficient to warn about this possible deadlock in the docs.

> you can still get a deadlock if you have a circular dependency of dbus calls that cycles back to this ObjectServer.

D-Bus calls are always more expensive and have a higher latency so using it for intra-process communication sounds like a bad idea in general. There are much better ways of doing that, e.g channels. So as you said, we shouldn't need to care about this case at all. Just document it, if really needed.

Comment (zeenix) [reply zbus_macros/src/lib.rs:225]: > > Since it's bounded, you could get this deadlock if you wait on any dbus recv operation inside your handler (the usual cause of such a wait would be "making a dbus call").

Actually looking back at the code, I'm not sure that's true. The ObjectServer's receiver is [for `METHOD_CALL` messages only](https://github.com/dbus2/zbus/blob/f45718536c7fba6b88c3db071bb90ccc2efa4522/zbus/src/connection/mod.rs#L957) so [any `METHDO_RETURN` messages won't be broadcasted to it](https://github.com/dbus2/zbus/blob/6a78d8ef191391c69c4fffe530741db67a314bac/zbus/src/connection/socket_reader.rs#L63) and hence socket reader task won't deadlock on broadcasting any method call returns while the method dispatch is waiting on a method call.

Did I miss something?

Comment (danieldg) [reply zbus_macros/src/lib.rs:225]: Yes: if parties on the bus have sent us `DEFAULT_MAX_QUEUED + 1` method calls, then `receive_msg` will block on the `broadcast_direct` call.  That prevents any new messages, regardless of type, from being processed.

Comment (zeenix) [reply zbus_macros/src/lib.rs:225]: > Yes: if parties on the bus have sent us `DEFAULT_MAX_QUEUED + 1` method calls, then `receive_msg` will block on the `broadcast_direct` call.

But why? The queues are separate. Perhaps it would help me understand if you could describe an example scenario.

Comment (danieldg) [reply zbus_macros/src/lib.rs:225]: 1. Client A makes the AccidentalDeadlockFunction method call; this ends in zbus's queue.
2. Client B sends `DEFAULT_MAX_QUEUED + 1` method calls to `GetFoo` (say, a badly designed API where you can get only one property at a time; this doesn't matter exactly, just the number of messages).
3. https://github.com/dbus2/zbus/blob/f45718536c7fba6b88c3db071bb90ccc2efa4522/zbus/src/connection/socket_reader.rs#L75 starts sending to the queue.  After `DEFAULT_MAX_QUEUED`, it blocks (2 messages left to send).
4. The ObjectServer dequeues one method call (ADF) and calls it. It makes a call to `DBusProxy::GetConnectionUnixProcessId`; this gets sent out.
5. The dbus server sends the reply on the socket.
6. `receive_msg` runs again, and is able to send one more message via `broadcast_direct`.  There's one `GetFoo` left (parsed but not sent), and just behind it is the reply from `GetConnectionUnixProcessId`.
7. Deadlock.

Comment (zeenix) [reply zbus_macros/src/lib.rs:225]: Thanks for explaining! Yeah, now I see it.

However, does having the option on the interface level eliminate this deadlock possibility? I think it only reduces the likelihood, assuming that developer disabling spawning, designs their interfaces and code carefully. If so, I don't see this as a good rationale for putting it on the interface level.

Comment (danieldg) [reply zbus_macros/src/lib.rs:225]: To avoid the deadlock, you could have the documentation declare: if you ever wait on this `Connection` object inside your `ObjectServer` (by making a method call without NoReply, etc), you must declare your interface to be spawning (or if the default is spawning, must not declare it no-spawn).  The `GetFoo` calls in this example could be declared no-spawn without any problems.

Putting it at the interface level is "closer" to the problem (you could even put it at the method level) and so it's more likely to be noticed - though you could argue that either way.  If you have *multiple* interfaces, making the entire connection no-spawn puts restrictions on all of them - which is more of a burden to check.

Comment (zeenix) [reply zbus_macros/src/lib.rs:225]: > Putting it at the interface level is "closer" to the problem (you could even put it at the method level)

Well, I think if we do this in a more granular way, method-level seems like the best option. If you hit this issue on an interface-level, there may not be something you could do about it (e.g if you're implementing a standardized interface, that you can not change).

@elmarco can you modify the PR to move the option to methods instead? I know this will require breaking changes on `Interface` but [as I wrote](https://github.com/dbus2/zbus/issues/674#issuecomment-2020310501) in #674, I think we should treat this an unstable API. We just need to document it more explicitly.

If so, please address other comments too and rebase the PR.

Comment (zeenix) [reply zbus_macros/src/lib.rs:225]: > I think we should treat this an unstable API. We just need to document it more explicitly.

Having said that, we should still try to make the change a bit more future-proof:

```rust

pub struct CallSettings {
    spawn_task: bool,
}

impl CallSettings {
    pub fn new() {
        MethodCallSettings::default()
    }

    pub fn set_spawn_task(mut self, spawn_task: bool) -> Self {
         self.spawn_task = spawn_task;

         self
    }
}
```

and then add an `Option<CallSettings>` to `Interface::{call, call_mut}`.

Comment (elmarco) [reply zbus_macros/src/lib.rs:225]: @zeenix if we are breaking Interface, I would rather implement part of my proposal from #674. Property and Method can have additional details, and there is no call & call_mut API duplication.

To me, the method granularity isn't the most appropriate. Usually, you have a related methods in an interface that you want to handle in order. Annotating each interface method is cumbersome, but that's fine to me if you believe it's better.

**Comment by zeenix** [comment]
@elmarco rebase please.

File: zbus/src/object_server/interface.rs
Comment (zeenix) [inline zbus/src/object_server/interface.rs:73]: * "with" -> "from".
* I think "out of order" isn't correct. The task are spawned in order, just that they may not be scheduled or the dispatched methods may not respond in order. Perhaps better add a separate note instead:

```
Note: When methods are called from separate tasks, they may not return in the order in which they were called. Typically, this is not a problem and therefore, the default implementation returns `true`.
```

That way, we also document the default value.

File: zbus/src/object_server/mod.rs
Comment (zeenix) [inline zbus/src/object_server/mod.rs:786]: * `read_lock` here is not a lock but a guard.
* I prefer naming the guards by what they contain (rather than naming it after their smart pointer container), so:

```rust
let with_spawn = {
     ...
     let iface = iface.read().await;
     ...
     iface.with_spawn()
};
```

This also gets rid of explicit `drop`. Call me a weirdo but I prefer using scopes over manual drops. :)

File: zbus/src/connection/mod.rs
Comment (zeenix) [inline zbus/src/connection/mod.rs:1031]: This is very confusing. So until the commit "♻️ zb: drop the mandatory task spawning", we were essentially spawning tasks twice for each method? So if we only reverted this commit, we'll be in a weird state.

Why not first do the needed refactor so that you don't need this additional commit?

**Review by zeenix — CHANGES_REQUESTED** [review]
LGTM in general.

File: zbus/src/connection/mod.rs
Comment (elmarco) [reply zbus/src/connection/mod.rs:1031]: Well, if we drop spawning before introducing it, we are in a weird state (worse since it regresses). What's wrong with spawning twice? Alternatively, just squash all changes, but that's harder to review/bisect.

File: zbus_macros/src/lib.rs
Comment (elmarco) [reply zbus_macros/src/lib.rs:223]: The default could be problematic. What do you suggest?

File: zbus_macros/src/lib.rs
Comment (elmarco) [reply zbus_macros/src/lib.rs:228]: Well, I don't like vague detailslike "in certain situations" either. but fine.

File: zbus/src/object_server/interface.rs
Comment (elmarco) [reply zbus/src/object_server/interface.rs:73]: It's not simply about "return in the order", it's also execution. I disagree about "Typically, this is not a problem" too.

File: zbus/src/object_server/mod.rs
Comment (elmarco) [reply zbus/src/object_server/mod.rs:786]: This is based on existing code, I just followed the same style. The dispatch usage requires different variables with explicit drop because of the read/write locking.

It can't be done the way you suggest.

File: zbus/src/connection/mod.rs
Comment (elmarco) [reply zbus/src/connection/mod.rs]: If you remove the lifetime cast, you get this compile error: ```argument requires that `msg` is borrowed for `'static```

The borrow-checker isn't able to correlate the lifetimes of 'a msg & hdr<'a>. They must become 'static to be given to spawn. 

Anyway, this is a temporary hack, as explained in the commit message.

File: zbus_macros/src/lib.rs
Comment (zeenix) [reply zbus_macros/src/lib.rs]: It's less about what I or you like, it's more about what's useful to the user. They don't need to know what internals are involved, just what they have to do, to avoid deadlocks.

File: zbus_macros/src/lib.rs
Comment (zeenix) [reply zbus_macros/src/lib.rs:223]: > The default could be problematic.

Only in some very rare specific use cases that assume unspecified requirements, so no need to scare most users who will never have any issues with it.

> What do you suggest?

"In certain situation, where clients assume unspecified requirements on services, this behavior could be problematic. For example, <your use case described here>. You can set this to `false` to enable compatibility with such clients."

File: zbus/src/connection/mod.rs
Comment (zeenix) [reply zbus/src/connection/mod.rs:1031]: whatever, I don't have the energy for debate this one. :)

File: zbus/src/object_server/mod.rs
Comment (zeenix) [reply zbus/src/object_server/mod.rs:786]: > The dispatch usage requires different variables with explicit drop because of the read/write locking.

Not sure, I follow. why can't my suggestion work? scoping is a very usual way to drop locks as soon as they're no longer needed.

Comment (zeenix) [reply zbus/src/object_server/mod.rs:786]: > This is based on existing code, I just followed the same style.

Pretty sure, the existing code doesn't call this boolean returning method, since you added it in this PR. :)

Comment (elmarco) [reply zbus/src/object_server/mod.rs]: ok, I added a scope to drop the explicit drop

File: zbus_macros/src/lib.rs
Comment (elmarco) [reply zbus_macros/src/lib.rs]: see new version

File: zbus/src/connection/mod.rs
Comment (zeenix) [inline zbus/src/connection/mod.rs:1012]: Seems this is now checked only **after** the spawn? This could mean that a misbehaving client can get us to spawn lots of tasks needlessly. Let's avoid that and check the message integrity etc before spawning the task?

**Review by zeenix — CHANGES_REQUESTED** [review]
LGTM otherwise. If you could, please do check if these changes somehow doesn't end up breaking busd (at some point we should just add busd to zbus' CI to ensure this).

File: zbus/src/connection/mod.rs
Comment (zeenix) [reply zbus/src/connection/mod.rs:1012]: * Member name is mandatory for method call messages so "unknown member" means an invalid message here.
* Regardless of this distinction, the point is that we should ensure we've all essentials in the method call before spawning a task for its dispatch.

Comment (elmarco) [reply zbus/src/connection/mod.rs:1012]: Because the way Interface::call{_mut} is designed, the return value needs to hold the lock. (in contrast with my proposal from https://github.com/dbus2/zbus/issues/674).

So if we want to avoid multiple redundant lookups, we need to do the Method resolution in the spawned task, we can't easily gather all the details before.

Comment (elmarco) [reply zbus/src/connection/mod.rs:1012]: (Msg and Header can become 'static,  but not 'a in Header<'a>)

Comment (zeenix) [reply zbus/src/connection/mod.rs:1012]: > But it fails the borrow-checker that it isn't able to match the lifetime of Msg and Header<'_> to 'static when moving those to a future.

Right, I guess because of `spawn` requirements. Still, as I said, I'd rather pay the small[1] penalty of header getter.

---

1. [As documented](https://docs.rs/zbus/latest/zbus/message/struct.Message.html#method.header), this method is no longer costly, just allocates a vec for the fields.

File: zbus/src/object_server/mod.rs
Comment (zeenix) [inline zbus/src/object_server/mod.rs:773]: * `FIXME: ` prefix would be nice.
* Feel free to make some changes already here, if needed ([I already made it clear that `Interface` is unstable API](https://github.com/dbus2/zbus/pull/711)). However I don't really think we need to change `Interface` here to be able to check everything upfront. The only/main part we really want to spawn off, is the user-provided handlers.
* This commit is only adding a check using the `hdr` we already have at hand so not sure `💩` is justified. :)

Comment (elmarco) [reply zbus/src/object_server/mod.rs:773]: 💩 : because we are doing multiple time the same check and the check is very limited: you can still spawn task if you provide a Member, even if it doesn't exist on the interface. Iow, I disagree this check is worthwhile.

File: zbus/src/object_server/mod.rs
Comment (elmarco) [reply zbus/src/object_server/mod.rs:794]: I can move the with_spawn to interface_lock(), that should get rid of the lock.

File: zbus/src/connection/mod.rs
Comment (zeenix) [reply zbus/src/connection/mod.rs:1012]: > I am not sure what you want to do.

I don't really see a problem with what you've already in your latest commit. So I'm as confused as you about what we're discussing here anymore. :)

File: zbus/src/object_server/mod.rs
Comment (zeenix) [reply zbus/src/object_server/mod.rs:773]: > It would be a major refactoring to move to the proposal I made.

Right, I didn't realize you mean you need **all** of your proposed changes here.

> The gain is great though, because there is no need to lock the interface to get those details: callback, spawn, introspection details etc.

No need to (even read) lock the interface to call method on it? How so? :thinking: Keep in mind that the trait can't be object safe if has associated constants or methods w/o receivers IIRC.

> because we are doing multiple time the same check and the check is very limited

We shouldn't need to check multiple times. Once it's established the the member is present, later code should just assume that it is and `unwrap()`.

> Iow, I disagree this check is worthwhile.

I explained already why I want it but more importantly, the check is already done in the existing code so spawning for messages w/o member present, would be a regression.

File: zbus/src/object_server/mod.rs
Comment (zeenix) [reply zbus/src/object_server/mod.rs:794]: How? :thinking: You still need to lock the interface's lock to make a call on it.

Comment (elmarco) [reply zbus/src/object_server/mod.rs:794]: To make a call, but not to gather rtti/metadata. that's the main point about #674, so it would be one step in this direction.

File: zbus/src/connection/mod.rs
Comment (elmarco) [reply zbus/src/connection/mod.rs:1012]: Can you do it on top? I feel stupid, I don't get what you want.

File: zbus/src/object_server/mod.rs
Comment (elmarco) [reply zbus/src/object_server/mod.rs:796]: squashed the 💩

File: zbus/src/object_server/mod.rs
Comment (elmarco) [reply zbus/src/object_server/mod.rs]: We don't need any locking to get the interface static RTTI. See "zb: gather some iface details earlier" patch for the idea. #674 should push this further, and allow to get rid of the read+write locks too.

Comment (zeenix) [reply zbus/src/object_server/mod.rs]: I'm not sure what you mean by RTTI and where I find this patch you're talking about but your proposed `Interface` trait in #674 uses associated constants and [hence can't be object-safe](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d838741ba246ea0d03f0031937ce1cf4) and we need object-safety cause we keep various interface impls in the same collection.

File: zbus/src/object_server/interface.rs
Comment (zeenix) [inline zbus/src/object_server/interface.rs:158]: Ah you got the info upfront and cached it, smart!

**Comment by zeenix** [comment]
@elmarco Ok to merge?

**Comment by zeenix** [comment]
Perhaps it would make sense to automatically disable spawning for method calls with `NO_REPLY_EXPECTED` flag set on them?

File: zbus/src/connection/builder.rs
Comment (zeenix) [inline zbus/src/connection/builder.rs:422]: best add an error variable for this. `Unsupported` seems misplaced since this is a valid failure and not some lack of support of a feature that should be supported.

**Review by zeenix — CHANGES_REQUESTED** [review]
LGTM otherwise.

File: zbus/src/connection/builder.rs
Comment (elmarco) [reply zbus/src/connection/builder.rs:422]: yeah, I wasn't so sure it is not worth to support. But maybe it can be changed anyway another day.

Comment (zeenix) [reply zbus/src/connection/builder.rs]: We both know that that day will likely never come. :) Let's just add a variant:

```rust
enum Error {
    // ...
    /// The given interface already exists at the given path. 
    InterfaceExists(InterfaceName<'static>, ObjectPath<'static>),
}
```

Not the best name or description, I know but I think it's still better than just throwing a generic error.

Comment (zeenix) [reply zbus/src/connection/builder.rs]: Our error struct is declared `non_exhaustive` so we should make use of that and added variants wherever needed.

**Review by zeenix — APPROVED** [review]
Good to go.

**Comment by zeenix** [comment]
So latest change is only the nitpick fix and a rebase on current main? If so, I'll merge.

**Comment by elmarco** [comment]
> So latest change is only the nitpick fix and a rebase on current main? If so, I'll merge.

yes, please merge

**Pull request merged**
